src/utils.ts

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import moment = require('moment');
+
+
+/*
+ * Copy the contents of one object to another, recursively.
+ *
+ * http://stackoverflow.com/questions/12317003/something-like-jquery-extend-but-standalone
+ */
+export
+function extend(target: any, source: any): any {
+  target = target || {};
+  for (var prop in source) {
+    if (typeof source[prop] === 'object') {
+      target[prop] = extend(target[prop], source[prop]);
+    } else {
+      target[prop] = source[prop];
+    }
+  }
+  return target;
+}
+
+
+/*
+ * Get a uuid as a string.
+ */
+export
+function uuid(): string {
+  /**
+   * http://www.ietf.org/rfc/rfc4122.txt
+   */
+  var s: string[] = [];
+  var hexDigits = "0123456789ABCDEF";
+  for (var i = 0; i < 32; i++) {
+    s[i] = hexDigits.substr(Math.floor(Math.random() * 0x10), 1);
+  }
+  s[12] = "4";  // bits 12-15 of the time_hi_and_version field to 0010
+  s[16] = hexDigits.substr((Number(s[16]) & 0x3) | 0x8, 1);  // bits 6-7 of the clock_seq_hi_and_reserved to 01
+
+  var uuid = s.join("");
+  return uuid;
+}
+
+
+/**
+ * Join a sequence of url components with '/'.
+ */
+export
+function urlPathJoin(...paths: string[]): string {
+
+  var url = '';
+  for (var i = 0; i < paths.length; i++) {
+    if (paths[i] === '') {
+      continue;
+    }
+    if (url.length > 0 && url[url.length - 1] != '/') {
+      url = url + '/' + paths[i];
+    } else {
+      url = url + paths[i];
+    }
+  }
+  url = url.replace(/\/\/+/, '/');
+  return url;
+}
+
+
+/**
+ * Encode just the components of a multi-segment uri,
+ * leaving '/' separators.
+ */
+export
+function encodeURIComponents(uri: string): string {
+  return uri.split('/').map(encodeURIComponent).join('/');
+}
+
+
+/**
+ * Join a sequence of url components with '/',
+ * encoding each component with encodeURIComponent.
+ */
+export
+function urlJoinEncode(...args: string[]): string {
+  return encodeURIComponents(urlPathJoin.apply(null, args));
+}
+
+
+// Properly detect the current browser.
+// http://stackoverflow.com/questions/2400935/browser-detection-in-javascript
+export
+var browser: string[] = (function() {
```

Comment:
you can use a typscript arrow function here: `= (() => { ... })();`, but I'm not going to die if you don't.